src/labels/point_placement.js

```diff
@@ -0,0 +1,137 @@
+// Logic for placing point labels along a line geometry
+
+import LabelPoint from './label_point';
+import {isCoordOutsideTile} from '../builders/common';
+
+const PLACEMENT = LabelPoint.PLACEMENT;
+const default_spacing = 50; // spacing of points along line in pixels
+
+export default function placePointsOnLine (line, size, options) {
+    let labels = [];
+    let strategy = options.placement;
+
+    switch (strategy){
+        case PLACEMENT.SPACED:
+            let result = getPositionsAndAngles(line, options);
+            // false will be returned if line have no length
+            if (!result) {
+                return [];
+            }
+
+            let positions = result.positions;
+            let angles = result.angles;
+            for (let i = 0; i < positions.length; i++){
+                let position = positions[i];
+                let angle = angles[i];
+                if (options.tile_edges === true || !isCoordOutsideTile(position)) {
+                    let label = new LabelPoint(position, size, options);
+                    label.angle = angle;
+                    labels.push(label);
+                }
+            }
+            break;
+        case PLACEMENT.VERTEX:
+            let p, q, label;
+            for (let i = 0; i < line.length - 1; i++){
+                p = line[i];
+                q = line[i + 1];
+                if (options.tile_edges === true || !isCoordOutsideTile(p)) {
+                    label = new LabelPoint(p, size, options);
+                    label.angle = getAngle(p, q, options.angle);
+                    labels.push(label);
+                }
+            }
+
+            // add last endpoint
+            label = new LabelPoint(q, size, options);
+            label.angle = getAngle(p, q, options.angle);
+            labels.push(label);
+            break;
+        case PLACEMENT.MIDPOINT:
+            for (let i = 0; i < line.length - 1; i++){
+                let p = line[i];
+                let q = line[i + 1];
+                let position = [
+                    0.5 * (p[0] + q[0]),
+                    0.5 * (p[1] + q[1])
+                ];
+                if (options.tile_edges === true || !isCoordOutsideTile(position)) {
+                    let label = new LabelPoint(position, size, options);
+                    label.angle = getAngle(p, q, options.angle);
+                    labels.push(label);
+                }
+            }
+            break;
+    }
+    return labels;
+}
+
+function getPositionsAndAngles(line, options){
+    let upp = options.units_per_pixel;
+    let spacing = (options.placement_spacing || default_spacing) * upp;
+
+    let length = getLineLength(line);
+
+    if (length === 0){
+        return false;
+    }
+
+    let num_labels = Math.max(Math.floor(length / spacing), 1);
+    let remainder = length - (num_labels - 1) * spacing;
+
+    let positions = [];
+    let angles = [];
+
+    let distance = 0.5 * remainder;
```

Comment:
@dmvaldman can you explain the heuristic of halving the remainder "placing space" here? I was thinking of starting the line interpolation from `remainder/num_labels`... Basically to distribute the actual placing between n labels. Will be good know if I am totally in the wrong direction here.

P.S. I am in middle of finishing this for ES. Just spacing case is left :P